File: leetcode/212-word-search-2/WordSearch.cpp
#include "bits/stdc++.h"
using namespace std;

void _print(const vector<string>& words) {
    for (const auto& s : words)
        cout << s << " ";
    cout << "\n";
}

const int R = 26;
const int dx[4] = {-1, 1, 0, 0};
const int dy[4] = {0, 0, -1, 1};

struct TrieNode {
    TrieNode* children[R];
    bool terminal;
    char c;

    TrieNode(char _c) : c(_c), terminal(false) {
        memset(children, 0, sizeof(children));
    }
};

class Trie {
private:
    TrieNode* root;

public:
    Trie() {
        root = new TrieNode('#');
    }

    Trie(const vector<string>& words) {
        root = new TrieNode('#');
        for (int i = 0; i < words.size(); ++i) {
            insert(words[i]);
        }
    }

    void insert(const string& s) {
        TrieNode *cur = root;
        for (int i = 0; i < s.size(); ++i) {
            int charIdx = s[i] - 'a';
            if (cur->children[charIdx] == nullptr) {
                cur->children[charIdx] = new TrieNode(s[i]);
            }
            cur = cur->children[charIdx];
        }
        cur->terminal = true;
    }

    bool isPrefix(const string& s) {
        TrieNode *cur = root;
        for (int i = 0; i < s.size(); ++i) {
            int charIdx = s[i] - 'a';
            if (cur->children[charIdx] == nullptr)
                return false;
            cur = cur->children[charIdx];
        }
        return true;
    }

    bool search(const string& s) {
        TrieNode *cur = root;
        for (int i = 0; i < s.size(); ++i) {
            int charIdx = s[i] - 'a';
            if (cur->children[charIdx] == nullptr)
                return false;
            cur = cur->children[charIdx];
        }
        return cur->terminal;
    }
};

void backtrack(int y, int x, string word, vector<vector<bool>>& visited, vector<vector<char>>& board, Trie& trie, set<string>& ret) {
    visited[y][x] = true;
    word += board[y][x];
    if (trie.isPrefix(word) == false) {
        visited[y][x] = false;
        return;
    }
    if (trie.search(word)) {
        ret.insert(word);
    }
    for (int k = 0; k < 4; ++k) {
        int ny = y + dy[k];
        int nx = x + dx[k];
        if (ny < 0 || ny >= board.size() || nx < 0 || nx >= board[0].size())
            continue;
        if (visited[ny][nx] == true)
            continue;
        backtrack(ny, nx, word, visited, board, trie, ret);
    }
    visited[y][x] = false;
}

vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
    Trie trie(words);
    set<string> ret;
    vector<vector<bool>> visited(board.size(), vector<bool>(board[0].size(), false));
    for (int i = 0; i < board.size(); ++i) {
        for (int j = 0; j < board[i].size(); ++j) {
            visited.assign(board.size(), vector<bool>(board[0].size(), false));
            backtrack(i, j, "", visited, board, trie, ret);
        }
    }
    return vector<string>(ret.begin(), ret.end());
}

int main(int argc, char const *argv[]) {
    {
        cout << "=====Example 1=====\n";
        vector<vector<char>> board{
            {'o','a','a','n'},
            {'e','t','a','e'},
            {'i','h','k','r'},
            {'i','f','l','v'},
        };
        vector<string> words{"oath", "pea", "eat", "rain"};
        vector<string> result = findWords(board, words);
        vector<string> correct_result{"eat", "oath"};
        cout << "Expected:\n\t";
        _print(correct_result);
        cout << "Output:\n\t";
        sort(result.begin(), result.end());
        _print(result);
    }
    {
        cout << "=====Example 2=====\n";
        vector<vector<char>> board{
            {'a', 'a'},
        };
        vector<string> words{"a"};
        vector<string> result = findWords(board, words);
        vector<string> correct_result{"a"};
        cout << "Expected:\n\t";
        _print(correct_result);
        cout << "Output:\n\t";
        sort(result.begin(), result.end());
        _print(result);
    }
    {
        cout << "=====Example 3=====\n";
        vector<vector<char>> board{
            {'a', 'b'}, {'c', 'd'}
        };
        vector<string> words{"acdb"};
        vector<string> result = findWords(board, words);
        vector<string> correct_result{"acdb"};
        cout << "Expected:\n\t";
        _print(correct_result);
        cout << "Output:\n\t";
        sort(result.begin(), result.end());
        _print(result);
    }
    return 0;
}

File: leetcode/79-word-search/word_search.py
# word_search.py
from typing import List


def exist(board: List[List[str]], word: str) -> bool:
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    rows = len(board)
    cols = len(board[0])

    def backtrack(i, y, x):
        if i == len(word):
            return True
        if y < 0 or y >= rows or x < 0 or x >= cols:
            return False
        if board[y][x] != word[i]:
            return False

        board[y][x] = "#"
        ret = False
        for k in range(len(dx)):
            ny = y + dy[k]
            nx = x + dx[k]
            if backtrack(i + 1, ny, nx):
                ret = True
                break
        board[y][x] = word[i]

        return ret

    for y in range(rows):
        for x in range(cols):
            if backtrack(0, y, x):
                return True

    return False


if __name__ == "__main__":
    board = [
        ["A", "B", "C", "E"],
        ["S", "F", "C", "S"],
        ["A", "D", "E", "E"],
    ]

    print("=====Example 1=====")
    word = "ABCCED"
    result = exist(board, word)
    correct_result = True
    print(f"Expected:\n\t{correct_result}")
    print(f"Output:\n\t{result}")

    print("=====Example 2=====")
    word = "SEE"
    result = exist(board, word)
    correct_result = True
    print(f"Expected:\n\t{correct_result}")
    print(f"Output:\n\t{result}")

    print("=====Example 3=====")
    word = "ABCB"
    result = exist(board, word)
    correct_result = False
    print(f"Expected:\n\t{correct_result}")
    print(f"Output:\n\t{result}")

    board = [["a"]]

    print("=====Example 4=====")
    word = "a"
    result = exist(board, word)
    correct_result = True
    print(f"Expected:\n\t{correct_result}")
    print(f"Output:\n\t{result}")
